### PR TITLE
TRDT-563

### PR DESCRIPTION
**Pull request by sam-miele**

### CODE REVIEW

File: app/controllers/IfsController.scala [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+import scala.concurrent.{ExecutionContext, Future}
+
+@Singleton()
+class IfsController @Inject()(headerValidator: HeaderValidatorAction,
+                              cc: ControllerComponents
+                             )(implicit ec: ExecutionContext)
+  extends BackendController(cc) with Logging {
+
+
+  val error1: Error = Error("INVALID_PAYLOAD", "Submission has not passed validation. Invalid payload.")
```
Comment (basotia) [inline app/controllers/IfsController.scala:36]: may be we can name after the errors.

File: test/controllers/IfsControllerSpec.scala [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+      eventName = "GenerateReport",
+      eventTimestamp = LocalDate.now(),
+      feedbackId = "feedbackId",
+      metadata = List(
+        Map("nino" -> "nino"),
+        Map("taxYear" -> "2023"),
+        Map("calculationId" -> calculationId),
+        Map("customerType" -> "Agent"),
+        Map("agentReferenceNumber" -> "12345"),
+        Map("calculationTimestamp" -> "2023-01-01T18:03:22")
```
Comment (basotia) [inline test/controllers/IfsControllerSpec.scala:48]: calculationTimeStamp should be in below format, we can refer code in api just to keep it uniform across
`   {
      "name": "calculationTimestamp",
      "value": "2019-02-15T09:35:15.094Z"
    }`

File: test/controllers/IfsControllerSpec.scala [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+      ),
+      payload = Some(IFRequestPayload(
+        messageId = "messageId",
+        englishAction = IFRequestPayloadAction(
+          title = "TODO - title",
```
Comment (basotia) [inline test/controllers/IfsControllerSpec.scala:53]: can we please remove todo, refer the response file under conf/response/submit. values from there can be used here, or just leave it a string without todo

File: test/controllers/IfsControllerSpec.scala [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        welshAction = IFRequestPayloadAction(
+          title = "TODO - title",
+          message = "TODO - message",
+          action = "TODO - action",
+          path = "TODO - path",
+          links = None
+        ))
+      ))
+
+    val fakeRequest: FakeRequest[JsValue] = FakeRequest("POST", "/interaction-data/store-interactions").withBody(Json.toJson(req)).withHeaders(("X-API-Key", "dummy-api-key"),("Content-Type", "application/json") )
```
Comment (basotia) [inline test/controllers/IfsControllerSpec.scala:68]: do we need api-key in this call?

File: test/controllers/IfsControllerSpec.scala [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    controller.submit().apply(fakeRequest)
+  }
+
+  "CipFraudController submitFraudInfo()" when {
```
Comment (basotia) [inline test/controllers/IfsControllerSpec.scala:72]: Controller name needs correction

File: test/controllers/IfsControllerSpec.scala [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+    "provided with a calculation id in metadata" must {
+      "return 204" in {
+        val result = submitStoreInteraction("good one")
```
Comment (basotia) [inline test/controllers/IfsControllerSpec.scala:76]: the calculationId values needs to be corrected as in this case it seems to be an invalid value, but we are returning no content, in below scenario for 400 , value seems to be valid

File: test/controllers/IfsControllerSpec.scala [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+        status(result) must be(NO_CONTENT)
+      }
+    }
+
+    "provided with a calculation id in metadata to trigger error 1" must {
+      "return 400" in {
+        val result = submitStoreInteraction("640090b4-06e3-4fef-a555-6fd0877dc7ca")
```
Comment (basotia) [inline test/controllers/IfsControllerSpec.scala:83]: pass invalid value as mentioned above

File: test/controllers/IfsControllerSpec.scala (hunk 4 above)
Comment (sam-miele) [reply test/controllers/IfsControllerSpec.scala:68]: I will double check.

File: test/controllers/IfsControllerSpec.scala (hunk 6 above)
Comment (sam-miele) [reply test/controllers/IfsControllerSpec.scala:76]: It was implemented so there are set values which will return an error, and anything outside of that will be OK. "No content" is the success branch. Similar to how the RDS stubs are implemented

File: test/controllers/IfsControllerSpec.scala (hunk 7 above)
Comment (sam-miele) [reply test/controllers/IfsControllerSpec.scala:83]: Same as above. If you'd like a static "success" value. We can agree on that, but we will need to decided on what to do for any other ids.

File: app/controllers/IfsController.scala [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+  val invalidCorrelationId: Error = Error("INVALID_CORRELATIONID", "Submission has not passed validation. Invalid header CorrelationId.")
+  val serviceUnavailable: Error = Error("SERVICE_UNAVAILABLE", "Dependent systems are currently not responding.")
+
+  def submit(): Action[JsValue] = {
+    headerValidator.async(parse.json) { implicit request: Request[JsValue] =>
+      Future {
+        request.body.validate[IFRequest] match {
+          case JsSuccess(value, _) => value.metadata.find(_.contains("calculationId")) match {
+            case Some(value) => value.get("calculationId") match {
+              case Some("404404b4-06e3-4fef-a555-6fd0877dc7ca") => BadRequest(Json.toJson(invalidCorrelationId))
```
Comment (basotia) [inline app/controllers/IfsController.scala:46]: Do we any other scenario where the request can be bad request? example missing feedbackId, fraudRiskHeaders? basically any mandatory data, thats not present we should also perform validation to see request is valid

**Review by basotia — DISMISSED** [review]

**Review by basotia — APPROVED** [review]

**Pull request merged**